### ISSUE TITLE
[DSR] - Debian 12 x64

### ISSUE THREAD

**Issue by Chouffy** [PowerShell/PowerShell#19889]
### Name of the Distribution

Debian

### Version of the Distribution

12

### Package Types

- [X] Deb
- [ ] RPM
- [ ] Tar.gz
- [ ] Snap - Stop! Please file your issue in [PowerShell-Snap](https://github.com/powershell/powershell-snap) instead.

### Processor Architecture

amd64

### .NET Core Support

- [X] The version and architecture of the Distribution is [supported by .NET Core](https://github.com/dotnet/core/blob/master/release-notes/5.0/5.0-supported-os.md#linux).

### Distribution Requirements

- [X] The version of the Distribution is supported for at least one year.
- [X] The version of the Distribution is not an [interim release](https://ubuntu.com/about/release-cycle) or equivalent.

### Progress

- [ ] An issue has been filed to create a Docker image in [PowerShell-Docker](https://github.com/powershell/powershell-docker)

### For PowerShell Team **ONLY**

- [ ] Docker image created
- [ ] Docker image published
- [ ] Distribution tested
- [ ] Update `packages.microsoft.com` deployment
- [ ] [Lifecycle](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/docs-conceptual/PowerShell-Support-Lifecycle.md) updated
- [ ] Documentation Updated

**Comment by Chouffy** [comment]
Note on possible workaround: It's possible to install Powershell on Debian 12 via Snap, but it's not ideal :)

**Comment by 237dmitry** [comment]
As a linux user I recommend to install from tarball (not dpkg, not snap (!!!) or another package manager). Unpack pwsh distribution to `/opt/pwsh` Make simlink of pwsh to the path. Set additional permission for Test-Connection cmdlet:

```powershell
sudo setcap cap_net_raw=eip /opt/pwsh/pwsh     # the real path to executable pwsh
```

You can change installation path as you want.

**Comment by kasini3000** [comment]
mark. i will update my script when here down.

https://gitee.com/chuanjiao10/kasini3000_agent_linux#debian12-kernel61-release-day-2023-06-17-

**Comment by tsugliani** [comment]
Very interested for this, as we are upgrading to Debian 12 (Bookworm)

Seems to be missing in the microsoft bookworm repo, available on debian 10/11

```bash
❯ cat /etc/apt/sources.list.d/microsoft.list
deb [arch=amd64 signed-by=/usr/share/keyrings/microsoft.gpg] https://packages.microsoft.com/repos/microsoft-debian-bookworm-prod bookworm main
```

**Comment by Erich-Chen** [comment]
It works by following [Microsoft installation guide ](https://learn.microsoft.com/en-us/powershell/scripting/install/install-debian?view=powershell-7.3) for bullseye, then manually editing file `/etc/apt/sources.list.d/microsoft.list` before install.

**Comment by prakashguru** [comment]
+1 for this. 

@TravisEz13 Will it possible for you to have a look at this? Something similar we have to do like debian 11 https://github.com/PowerShell/PowerShell/issues/10923

**Comment by ExNought** [comment]
I am biting nails to get PowerShell on Debian 12. This is affecting my work since I am already on Debian 12 and VSC does not recognize PowerShell anymore. I currently use PowerShell from the legacy sources.

**Comment by rhubarb-geek-nz** [comment]
What is the issue with downloading the Debian package directly from the releases page?

https://github.com/PowerShell/PowerShell/releases/download/v7.4.0/powershell_7.4.0-1.deb_amd64.deb

**Comment by eliminmax** [comment]
> What is the issue with downloading the Debian package directly from the releases page?
> 
> https://github.com/PowerShell/PowerShell/releases/download/v7.4.0/powershell_7.4.0-1.deb_amd64.deb

It won't get updated with APT that way, and there's no guarantee that it will work on any particular version of Debian.

**Comment by prakashguru** [comment]
Dear @eliminmax I dont think there will be an issue to download manually the powershell and install it on my machine. 

But some of our customers using our application which has powershell as a dependency unable to install it on debian 12 because powershell is not published for debian 12. 

There are heavily restricted environments where installing packages outside of the official repositories are prohibited. 

So it is important and required to be published!

**Comment by AndrewSav** [comment]
@rhubarb-geek-nz  for anyone who had ever worked in an enterprise environment it is obvious why this would not be very manageable. MS start to provide those for a reason, and for the same reason we ask them to continue to do so.

**Comment by rhubarb-geek-nz** [comment]
> @rhubarb-geek-nz for anyone who had ever worked in an enterprise environment it is obvious why this would not be very manageable. MS start to provide those for a reason, and for the same reason we ask them to continue to do so.

Absolutely true but it is not a show stopper. The published debian file does work and it is not a rebuild required, just an addition to the packages.microsoft.com. If you are in an enterprise environment you may be hosting your own APT repositories or using downloads directly with curl when creating docker images.

If you check the recent debian packages on packages.microsoft.com they are the same file

```
#!/bin/sh -e

for o in debian ubuntu
do
        for v in 10 11 12 18.04 18.10 19.04 19.20 20.04 20.10 21.04 21.10 22.04 22.10 23.04 23.10
        do
                if curl https://packages.microsoft.com/$o/$v/prod/pool/main/p/powershell/powershell_7.4.0-1.deb_amd64.deb --output powershell_7.4.0-1.$o.$v.deb_amd64.deb --location --fail --silent
                then
                        sha256sum powershell_7.4.0-1.$o.$v.deb_amd64.deb
                fi
        done
done
```

```
550331e243a7311e1bf0ca83d8c96c5e78ba82a765c4a9ba34aa0db6eb9a9f22  powershell_7.4.0-1.debian.10.deb_amd64.deb
550331e243a7311e1bf0ca83d8c96c5e78ba82a765c4a9ba34aa0db6eb9a9f22  powershell_7.4.0-1.debian.11.deb_amd64.deb
550331e243a7311e1bf0ca83d8c96c5e78ba82a765c4a9ba34aa0db6eb9a9f22  powershell_7.4.0-1.ubuntu.18.04.deb_amd64.deb
550331e243a7311e1bf0ca83d8c96c5e78ba82a765c4a9ba34aa0db6eb9a9f22  powershell_7.4.0-1.ubuntu.20.04.deb_amd64.deb
550331e243a7311e1bf0ca83d8c96c5e78ba82a765c4a9ba34aa0db6eb9a9f22  powershell_7.4.0-1.ubuntu.22.04.deb_amd64.deb
```

The packages.microsoft.com repository for Ubuntu 22.04 amd64 is problematic because the dotnet-sdk-6.0 conflict with those distributed by ubuntu themselves. They have different version suffixes and install in different directories.

I have asked for a couple of years for armhf and aarch64 debian files but them not existing does not prevent me creating my own [APT repository](https://packages.rhubarb.geek.nz/debian/12/pool/main/p/powershell/) with repackaged files for armhf/aarch64.

**Comment by prakashguru** [comment]
@rhubarb-geek-nz No it is a show stopper!

We do have an apt repo. Technically I can host the powershell.deb package into it and distribute. But practically I can't due to the following reasons, 

- Have to go through license, compliance and security review to host a third party package into my company repository. That will be a nightmare.
- Security updates! Whenever MS releases security updates, I have to keep track of those and constantly update and again go through those reviews. 

So practically it is a no no.

**Comment by rhubarb-geek-nz** [comment]
> @rhubarb-geek-nz No it is a show stopper!

That is your call, you do have the existing work around mechanisms but you chose not to use them.

**Comment by Daverd-GM** [comment]
It would be nice to add the PowerShell.deb versions to the https://packages.microsoft.com/debian/12/prod/
for it to be available in the [Extrepo-Data]( https://salsa.debian.org/extrepo-team/extrepo-data) Project, which is the Debian tool for managing [external repositories] (https://salsa.debian.org/extrepo-team/extrepo) from the command line

**Comment by nanonyme** [comment]
It's not really the delay itself. Debian makes releases fairly reliably every two years and it's frustrating customers have to keep creating these tickets rather than it being baked into Powershell maintenance process.

**Comment by nanonyme** [comment]
Please don't get me wrong, I understand Microsoft has processes here where apparently supporting new Debian releases are tied to PowerShell releases which makes things awkward. Regardless, Debian releases are reasonably predictable in when they happen and there should be nothing preventing releasing PowerShell even *before* Debian actual release date if it's convenient.

**Comment by eliminmax** [comment]
> Please don't get me wrong, I understand Microsoft has processes here where apparently supporting new Debian releases are tied to PowerShell releases which makes things awkward. Regardless, Debian releases are reasonably predictable in when they happen and there should be nothing preventing releasing PowerShell even _before_ Debian actual release date if it's convenient.

I was almost done opening a separate issue asking them to do that, but then I couldn't figure out a way to programmatically check if the latest Debian release hit its soft freeze period, so I am not sure if it's as simple as I'd assumed, and I'm sure the PowerShell maintainers have a lot of other responsibilities as is. It would be nice if they'd have been quicker about addressing this issue, but, while new Debian releases are every odd-numbered years, sometime in the Summer, that's not a hard rule, so I don't know how simple it would be to do this.

**Comment by rhubarb-geek-nz** [comment]
I have seen that the debian deb file is actually the same one used for multiple OS, [sha256sum](https://github.com/PowerShell/PowerShell/issues/19889#issuecomment-1826274419), it is not a unique build per platform.

I hope it appears in packages.microsoft.com only after it has passed regression testing on each of those platforms.

**Comment by theJasonHelmick** [comment]
Hi Folks - I appreciate your patience.  This has bene moved from our backlog to planned in the next couple of months. I don;t have an exact timeline yet as we work to improve our distro pipeline.

**Comment by Gonkers** [comment]
Thanks for letting us know this is now on the list. I was quite confused when the apt sources didn't work until I came across this thread in search for a solution.

**Comment by nanonyme** [comment]
@theJasonHelmick given it's now around three months from your comment. Can you confirm if this is reaching top of backlog?

**Comment by theJasonHelmick** [comment]
Hi @nanonyme  - Yes, this is priority 1 on our backlog. We are working through some new release pipelines and will update this issue and the releases when complete. My current estimate is late July.

**Comment by OronDF343** [comment]
Any updates?

This issue is one of the top reasons that I always prefer bash over PowerShell, even in Azure Pipelines.

**Comment by nanonyme** [comment]
@theJasonHelmick we would please urgently need this. Jenkins has dropped support for OpenJDK in Debian 11 and upgrade to Debian 12 needs to happen asap. This has now been the blocker for us for last half a year.

**Comment by CodeMystic21** [comment]
Fuck Microsoft and their shit Powershell.
Just use bash.
Or any scripting language like python, go, ...

**Comment by nanonyme** [comment]
@CodeMystic21 let's please keep language civil to avoid the unfortunate outcome the entire issue gets locked for externals.

**Comment by MeikelLP** [comment]
> Fuck Microsoft and their shit Powershell. Just use bash. Or any scripting language like python, go, ...

"scripting language like go" 👀

**Comment by nanonyme** [comment]
@MeikelLP please don't quote reply. If Microsoft deletes the comment later, it will still be visible in your quote.

**Comment by theJasonHelmick** [comment]
I appreciate the frustration - however please stick to our code of conduct.  We are planning to begin to release new bistros at the beginning of September.  These releases will begin with Ubuntu and Azure Linux 3, followed by Debian.  We are working to release these as quickly as possible while minting security and quality.  As we begin shipping, I will notify and close in these issues, and you track planned/backlog releases here: https://github.com/orgs/PowerShell/projects/23

**Comment by kallisti5** [comment]
>  September.. These releases will begin with Ubuntu and Azure Linux 3, followed by Debian.

I do want to add that comments like this are frustrating, and show some level of disconnection at Microsoft about how open source works.  The current pace means Microsoft will be releasing powershell for versions of Debian less than 6 months from the next major version of Debian (13, early 2025)


I personally think the engineers involved need to take a hard look at what processes they're following to create an environment where getting a powershell release artifact built for an operating system 1.5 years after its release is a hard lift.

Open source teams of 1-2 people working for their own hobby enjoyment can update their software for new releases within a few months... Why can't a team with seemingly infinite resources do it within 2 years?

**Comment by nanonyme** [comment]
> Open source teams of 1-2 people working for their own hobby enjoyment can update their software for new releases within a few months... Why can't a team with seemingly infinite resources do it within 2 years?

Not part of Microsoft but my impression is there's some level of big-corporation syndrome. You look at Microsoft and see massive corporation with over 100 000 developers, people inside look at a bunch of small teams for project X, others for project Y etc. Microsoft doesn't normally support Debian until after release (you can take a look at prior similar issues) so it's not really two years. Debian 12 was released June 2023. I would hope some retrospective to happen though on how to improve the process. 15 months is a pretty long time to support a major distro which is widely used as build server.

**Comment by theJasonHelmick** [comment]
This distribution has published. https://mcr.microsoft.com/product/powershell/tags

**Comment by masterflitzer** [comment]
am i missing something?

i cannot find it in the debian 12 microsoft repo: <https://packages.microsoft.com/debian/12/prod/pool/main/p/>

i would've expected it at this location: <https://packages.microsoft.com/debian/12/prod/pool/main/p/powershell/>,
similar to debian 11: <https://packages.microsoft.com/debian/11/prod/pool/main/p/powershell/>

**Comment by kallisti5** [comment]
It looks like someone at Microsoft doesn't understand Docker images != Debian apt repositories :laughing:   Ow my lungs hurt from laughing.

**Comment by rafaelantony** [comment]
@theJasonHelmick, please take a look at @masterflitzer and @kallisti5 comments above.
We still need an apt repository for Debian 12!

**Comment by prakashguru** [comment]
> This distribution has published. https://mcr.microsoft.com/product/powershell/tags
> 
> > **[Microsoft Artifact Registry](https://mcr.microsoft.com/product/powershell/tags)**Microsoft Artifact Registry (also known as Microsoft Container Registry or MCR) Discovery Portal

That is the incorrect location to publish the deb file. It should be in the debian 12 microsoft repo. [https://packages.microsoft.com/debian/12/prod](https://packages.microsoft.com/debian/12/prod/pool/main/p/)

**Comment by GABRIELNGBTUC** [comment]
It's hard to understand what is preventing them to add the deb files to the microsoft repository.

Ok you tested the package on debian 12 and created a docker image (that no one on this issue asked for). 

Now that it is tested, there shouldn't be any blockers left to just drop the deb file into the package repository. 

I'm sure no one here would even mind if it was marked as an "unstable" package or something. Many have been using the direct method install because of this issue and didn't encounter any problem. If it's a case of "we don't want to support it", still give us the option to use the package repository but explicitly say so that support will be non-existent / best-effort.

**Comment by masterflitzer** [comment]
opened one over here: #24334

**Comment by theJasonHelmick** [comment]
Hey Folks -- thanks for the feedback. We are finishing some internal pipeline work and will post the deb files to Packages - PMC - shortly.  (hopefully this week)

**Comment by quaeryth** [comment]
Looking forward to seeing this resolved soon. I'd also like to second the concern in part of @kallisti5's comment:

> The current pace means Microsoft will be releasing powershell for versions of Debian less than 6 months from the next major version of Debian (13, early 2025)

**Comment by nanonyme** [comment]
> Looking forward to seeing this resolved soon. I'd also like to second the concern in part of @kallisti5's comment:

Debian has statistically usually released during Summer. It's around half a year of stabilization from the point new release gets frozen.

**Comment by xenadmin** [comment]
@theJasonHelmick Any news on this? 
We are now roughly ~9 months away from the release of Debian 13 and we urgently need PowerShell 7 support on Debian 12.

**Comment by kallisti5** [comment]
I just double checked, Vegas odds on Powershell deb's getting released before Debian 13 is out are `-5.2`

**Comment by AndrewSav** [comment]
March 16 - "in the next couple of months"
June 5 - "this is priority 1 on our backlog"
August 24 - "planning to begin to release new distros at the beginning of September"
September 18 - "this distribution has published" - it was not and the issue was closed
Sepetmber 24 - "hopefully this week" - the issue remains closed
October 2 - the issue is reopened

Two weeks after the last communication there is still nothing. My prediction: they will soon realise that it's not feasible making any commitments and then looking like this, so they will keep silent in the future. It's much easier to ignore, does not make any worse of an impression than it has already and good for one's health.

> The current pace means Microsoft will be releasing powershell for versions of Debian less than 6 months from the next major version of Debian (13, early 2025)

I wish I could excercise optimism with such abandon!

**Comment by 24mu13** [comment]
What's about to maintain the Debian package in their ecosystem?
https://www.debian.org/doc/manuals/debmake-doc/index.en.html

Do you think is feasible from licensing perspective?

**Comment by nanonyme** [comment]
> Hey Folks -- thanks for the feedback. We are finishing some internal pipeline work and will post the deb files to Packages - PMC - shortly. (hopefully this week)

Hey, would it be possible to provide any visibility over the work? If there are blockers, it would be useful to know exactly what and see their resolution's progress. Currently for us outsiders it looks very much like nothing's happening.

**Comment by mbentley** [comment]
> What's about to maintain the Debian package in their ecosystem? https://www.debian.org/doc/manuals/debmake-doc/index.en.html
> 
> Do you think is feasible from licensing perspective?
> 
> > **[Guide for Debian Maintainers](https://www.debian.org/doc/manuals/debmake-doc/index.en.html)**

That would likely increase the difficulty of maintenance for them and not really add any value; plus Debian has different support lifecycles than what I would guess that Microsoft *likely* would want to follow.  When you have software that updates pretty frequently and may evolve quickly, having their own package repo would allow them to more easily get packages for new versions available in a wider number of distros and not have to go through backports.

**Comment by 24mu13** [comment]
> > What's about to maintain the Debian package in their ecosystem? https://www.debian.org/doc/manuals/debmake-doc/index.en.html
> > Do you think is feasible from licensing perspective?
> > > **[Guide for Debian Maintainers](https://www.debian.org/doc/manuals/debmake-doc/index.en.html)**
> 
> That would likely increase the difficulty of maintenance for them and not really add any value; plus Debian has different support lifecycles than what I would guess that Microsoft _likely_ would want to follow. When you have software that updates pretty frequently and may evolve quickly, having their own package repo would allow them to more easily get packages for new versions available in a wider number of distros and not have to go through backports.
> 
> > **[Guide for Debian Maintainers](https://www.debian.org/doc/manuals/debmake-doc/index.en.html)**

I meant the community (us) would maintain it, not Microsoft; we saw in fact that is not able neither to keep the very slow Debian lifecycle...

**Comment by rhubarb-geek-nz** [comment]
I am happy that Ubuntu repackages the dotnet-sdk and provides regular updates. I have asked Canonical to [Add PowerShell to APT repository](https://bugs.launchpad.net/ubuntu/+bug/2080715)

Given Debian is a base for so many distributions and its pace, I don't think it is appropriate for either dotnet-sdk or PowerShell. The packages.microsoft.com generally does a good job providing up to date packages for Debian.

**Comment by 24mu13** [comment]
> I am happy that Ubuntu repackages the dotnet-sdk and provides regular updates. I have asked Canonical to [Add PowerShell to APT repository](https://bugs.launchpad.net/ubuntu/+bug/2080715)
> 
> Given Debian is a base for so many distributions and its pace, I don't think it is appropriate for either dotnet-sdk or PowerShell. The packages.microsoft.com generally does a good job providing up to date packages for Debian.

I don't get the point: are you happy with Ubuntu packages or Microsoft repository?

**Comment by rhubarb-geek-nz** [comment]
> I don't get the point: are you happy with Ubuntu packages or Microsoft repository?

I am happy with packages.microsoft.com for Debian. But for Ubuntu, given it has already diverged by hosting dotnet-sdk and this project is very reluctant to produce a working Ubuntu package for 24.04, I suggest Ubuntu would be better for Ubuntu.

**Comment by 24mu13** [comment]
We've got the package finally! Thank you Microsoft

**Comment by theJasonHelmick** [comment]
This package is now available at PMC. For more information see the PMC Readme:  
[https://packages.microsoft.com/docs/readme.txt](https://packages.microsoft.com/docs/readme.txt)   
  
We are still working on the docker image pipeline. I will post an update when we have published  
images to the MCR.

**Comment by theJasonHelmick** [comment]
The Debian 12 distribution is released to PMC and MCR.
